test/gtest/ucp/test_ucp_atomic.cc

```diff
@@ -153,17 +201,60 @@ class test_ucp_atomic : public test_ucp_memheap {
         return request_wait(status_ptr);
     }
 
+    void test_mem_types(send_func_t send_func, unsigned num_iters,
+                        uint64_t op_mask, int is_ep_flush) {
+        int atomic_mode = get_variant_value() & ATOMIC_MODE;
+        std::vector<std::vector<ucs_memory_type_t> > pairs =
+                ucs::supported_mem_type_pairs();
+
+        for (size_t i = 0; i < pairs.size(); ++i) {
+            if (!UCP_MEM_IS_HOST(pairs[i][0]) || !UCP_MEM_IS_HOST(pairs[i][1])) {
+                /* Memory type atomics are fully supported only with new protocols */
+                if (!(get_variant_value() & ENABLE_PROTO)) {
+                    continue;
+                }
+
+                static const std::string tls[] = { "ud_v", "ud_x", "rc_v", "tcp" };
+                /* Memory type atomics emulation not supported yet */
+                if (atomic_mode == UCP_ATOMIC_MODE_CPU ||
```

Comment:
if source is GPU and destination is host - can still use SW protocol (sender will use cudaMemcpy/gdrcopy)